promise.go

```diff
@@ -0,0 +1,111 @@
+package cache
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+type Promise[T any] struct {
+	done  <-chan struct{}
+	value T
+	err   error
+}
+
+func NewPromise[T any]() (*Promise[T], func(T, error)) {
+	ch := make(chan struct{})
+	p := &Promise[T]{
+		done: ch,
+	}
+	resolve := func(value T, err error) {
+		p.value = value
+		p.err = err
+		close(ch)
+	}
+	return p, resolve
+}
+
+type inFlightPromise[T any] struct {
+	*Promise[T]
+	waiting atomic.Int64
+}
+
+type CachedResourceCoalescing[K comparable, T any] struct {
+	OnErr    func(error)
+	cache    SimpleCache
+	cacheMX  sync.RWMutex
+	inFlight map[K]*inFlightPromise[T]
+	once     sync.Once
+}
+
+func NewCachedResourceCoalescing[K comparable, T any](cache SimpleCache) *CachedResourceCoalescing[K, T] {
+	return &CachedResourceCoalescing[K, T]{
+		cache:    cache,
+		inFlight: make(map[K]*inFlightPromise[T]),
+	}
+}
+
+// GetOnceWithStore is a helper function that fetches a value from the cache or executes the provided function to get the value.
+func GetOnceWithStore[K comparable, T any](
+	ctx context.Context,
+	crc *CachedResourceCoalescing[K, T],
+	key K,
+	get func() (result T, storeInCache bool, err error),
+) (result T, err error) {
+	var cachedRes T
+	cacheErr := crc.cache.Get(ctx, key, &cachedRes)
+	if cacheErr == nil {
+		return cachedRes, nil
+	}
+
+	crc.cacheMX.Lock()
+	promise, found := crc.inFlight[key]
+	if found {
+		promise.waiting.Add(1)
+		defer promise.waiting.Add(-1)
+		crc.cacheMX.Unlock()
+		select {
+		case <-ctx.Done():
+			return result, ctx.Err()
+		case <-promise.done:
+			return promise.value, promise.err
```

Comment:
without default we could have double unlock?

Reply:
Nah, without a default case, the select would end up only on one of those channels. While we don't receive any value, it stays locked.

Look at this go snippet: [go playground](https://go.dev/play/p/_ckz1Vqyt8K)
it only jumps out of the select after one of those timers finishes.